bpf/lib/nodeport.h

```diff
@@ -522,6 +522,87 @@ static __always_inline int nodeport_nat_ipv4_rev(struct __sk_buff *skb,
 	return NODEPORT_DO_NAT_IPV4(addr, NAT_DIR_INGRESS);
 }
 
+/* Helper function to set the IPv4 option for DSR when a backend is remote.
+ * NOTE: Revalidate data after calling the function.
+ */
+static inline int set_dsr_opt4(struct __sk_buff *skb, struct iphdr *ip4,
+			       __be32 svc_addr, __be32 svc_port)
+{
+	union tcp_flags tcp_flags = { .value = 0 };
+
+	if (ip4->protocol == IPPROTO_TCP) {
+		if (skb_load_bytes(skb, ETH_HLEN + sizeof(*ip4) + 12,
+				   &tcp_flags, 2) < 0)
+			return DROP_CT_INVALID_HDR;
+		// Setting the option is required only for the first packet
+		// (SYN), in the case of TCP, as for further packets of the
+		// same connection a remote node will use a NAT entry to
+		// reverse xlate a reply.
+		if (!(tcp_flags.value & (TCP_FLAG_SYN)))
+			return 0;
+	}
+
+	ip4->ihl += 0x2; // To accommodate u64 option
+	ip4->tot_len = bpf_htons(bpf_ntohs(ip4->tot_len) + 0x8);
```

Comment:
the `0x2` is to accommodate u64 option, and what is the `0x8` for?

Reply:
`ihl` counts in 32-bit words, and stores IPv4 header len. `tot_len` counts in bytes, and stores IPv4 header + header len.